tools/pythonpkg/duckdb/duckdb_query_graph/__init__.py

```diff
@@ -0,0 +1,285 @@
+import json
+import os
+from functools import reduce
+
+qgraph_css = """
+.styled-table {
+	border-collapse: collapse;
+	margin: 25px 0;
+	font-size: 0.9em;
+	font-family: sans-serif;
+	min-width: 400px;
+	box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);
+}
+.styled-table thead tr {
+	background-color: #009879;
+	color: #ffffff;
+	text-align: left;
+}
+.styled-table th,
+.styled-table td {
+	padding: 12px 15px;
+}
+.styled-table tbody tr {
+	border-bottom: 1px solid #dddddd;
+}
+
+.styled-table tbody tr:nth-of-type(even) {
+	background-color: #f3f3f3;
+}
+
+.styled-table tbody tr:last-of-type {
+	border-bottom: 2px solid #009879;
+}
+
+.node-body {
+	font-size:15px;
+}
+.tf-nc {
+	position: relative;
+	width: 250px;
+	text-align: center;
+	background-color: #fff100;
+}
+"""
+
+
+class NodeTiming:
+
+	def __init__(self, phase, time):
+		self.phase = phase
+		self.time = time
+		# percentage is determined later.
+		self.percentage = 0
+
+	def calculate_percentage(self, total_time):
+		self.percentage = self.time/total_time
+
+	def combine_timing(l, r):
+		# TODO: can only add timings for same-phase nodes
+		total_time = l.time + r.time
+		return NodeTiming(l.phase, total_time)
+
+class AllTimings:
+
+	def __init__(self):
+		self.phase_to_timings = {}
+
+	def add_node_timing(self, node_timing):
+		if node_timing.phase in self.phase_to_timings:
+			self.phase_to_timings[node_timing.phase].append(node_timing)
+			return
+		self.phase_to_timings[node_timing.phase] = [node_timing]
+
+	def get_phase_timings(self, phase):
+		return self.phase_to_timings[phase]
+
+	def get_summary_phase_timings(self, phase):
+		return reduce(NodeTiming.combine_timing, self.phase_to_timings[phase])
+
+	def get_phases(self):
+		phases = list(self.phase_to_timings.keys())
+		phases.sort(key=lambda x: (self.get_summary_phase_timings(x)).time)
+		phases.reverse()
+		return phases
+
+	def get_sum_of_all_timings(self):
+		total_timing_sum = 0
+		for phase in self.phase_to_timings.keys():
+			total_timing_sum += self.get_summary_phase_timings(phase).time
+		return total_timing_sum
+
+QUERY_TIMINGS = AllTimings()
```

Comment:
I don't see this being reset anywhere, can the graph generator only be used once per interpreter session?

Reply:
That's the idea yea, but it's probably good to just reset this every time a new profile output is passed

Reply:
Better to not use a global to begin with probably?